### PR TITLE
Added Workspace Menu

### PR DESCRIPTION
**Pull request by mananjadhav**
cc - @chiragsalian 

### Details

- Adds a generic popup menu to Header with options
- Uses this popup menu to add "New workspace" and "Delete Workspace" options

### Fixed Issues

$ https://github.com/Expensify/App/issues/5185

### Tests
1. Tested the popup for all platforms - including hiding close button on desktop
2. Checked the options for New Workspace
3. Changed by updating language and translations
4. Pending: Delete Workspace Testing

### QA Steps
1. Go to the right sidebar
2. Click on any Workspace item
3. You should see three dots (more options) menu, hover over it
4. You should see two options - New Workspace and Delete Workspace
5. Test both the actions are working fine
6. Try the same steps for mobile, except that the popup should be docked at the bottom

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

web-workspace-menu

#### Mobile Web

mweb-workspace-menu

#### Desktop

desktop-workspace-menu

#### iOS

ios-workspace-menu

#### Android

android-workspace-menu

### CODE REVIEW

**Comment by mananjadhav** [comment]
Hi @chiragsalian,
I've got the whole PR ready except for the delete API call.

One major question is, I was trying to get the latest pull and it turns out I am not able to access this component. Did we change the flow? 

image

This is the sidebar I see on click. Does the component have to be moved to this one?

**Comment by mananjadhav** [comment]
@chiragsalian  Were you able to check the questions in the previous comment ?

**Comment by mananjadhav** [comment]
@chiragsalian I've made updates to the PR and currently testing. Can you help with one last thing? 

```
return Onyx.set(`${ONYXKEYS.COLLECTION.POLICY}${response.policyID}`, null);
```

I am using the above code to unset the value for the policy.  How do I delete a key from Onyx?

**Comment by chiragsalian** [comment]
So there is no way to delete a key from Onyx atm. We usually just set it to empty string or null just as you have.
So your code looks good. The only thing i would say is that i don't think request returns a policyID so that code of yours should most likely be,
```
return Onyx.set(`${ONYXKEYS.COLLECTION.POLICY}${policyID}`, null);
```
Once you update that just make sure to test that setting it to null doesn't break anything else 🙂

**Comment by mananjadhav** [comment]
@chiragsalian finally the PR complete but for some reason, E2E iOS Tests failed.

**Comment by mananjadhav** [comment]
@chiragsalian Reminder.

I'll resolve the conflicts once the review is complete. `WorkspaceInitialPage` is a functional component so even the smallest change is conflicting and I have to fix it all over again.

**Comment by chiragsalian** [comment]
Sorry for the delay. Last week i was really busy with other tasks and then i was on vacation for a few days. Now I'm back 100%. Looking into this. Hmm, the E2E test failure looks odd. As a first step, whenever there is an unusual travis test failure first re-run the test, and then update branch with main branch (you can do just the latter if you want since it will trigger travis test). Anyway i've re-run the job. Let's see if it fails at the same spot. I'll review the code in the meantime.

File: src/libs/actions/Policy.js [hunk 1]
```diff
@@ -131,6 +131,41 @@ function createAndNavigate(name = '') {
...
+
+/**
+ * Delete the policy
+ *
+ * @param {String} [policyID]
+ * @param {Boolean} shouldGrowl
+ * @param {Boolean} shouldAutomaticallyReroute
+ * @returns {Promise}
+ */
+function deletePolicy(policyID = '', shouldGrowl = true, shouldAutomaticallyReroute = true) {
```
Comment (chiragsalian) [inline src/libs/actions/Policy.js:143]: Why is policyID optional? It shouldn't be since the backend needs a policyID to know what to delete.

Comment (chiragsalian) [reply src/libs/actions/Policy.js:143]: I don't see shouldGrowl and shouldAutomaticallyReroute passed when we call this method. If that's the case let's not preoptimize here and remove these two params from the method call and just do the growl and reroute in the method. In future if there is a scenario to not show growl or reroute then we'll update as needed.

File: src/libs/actions/Policy.js [hunk 2]
```diff
@@ -131,6 +131,41 @@ function createAndNavigate(name = '') {
...
+                Growl.error(errorMessage, 5000);
+                return;
+            }
+
+            if (shouldGrowl) {
+                Growl.show(translateLocal('workspace.common.growlMessageOnDelete'), CONST.GROWL.SUCCESS, 3000);
+            }
+
+            // Removing the workspace data from Onyx as well
+            return Onyx.set(`${ONYXKEYS.COLLECTION.POLICY}${policyID}`, null);
```
Comment (chiragsalian) [inline src/libs/actions/Policy.js:158]: i believe you are missing the underscore `_` between your policy and policyID thats why the actual onyx variable is not getting cleared which is why i can still see the workspace after deleting it.

**Review by chiragsalian — CHANGES_REQUESTED** [review]
Tested the PR. I'm facing a few issues,
1. The moment a new workspace is created the URL show the reportID i.e., http://localhost:8080/r/255 for me instead of the workspace policyID url i.e., http://localhost:8080/workspace/52E30B06BC2CE462 (the only way to get the correct URL is by closing the workspace menu and opening it again which is weird)
2. Creating a new workspace ends up in a loop of tons of same network requests. Let me know if you notice the same as well.

**Comment by mananjadhav** [comment]
@chiragsalian: PR Updated with the changes. About the two issues:

1. It seems to be happening due to `Navigation.dismissModal()` in `Policy.js:navigateToPolicy` method. If I remove the `dismissModal` call, it works fine. I am not sure why it was added. Can you confirm?

2. I had this issue when I raised the first draft, but not later when I submitted with screenshots. Tried creating multiple workspaces and it didn't end into a loop. Can you take the latest pull and check?

**Comment by mananjadhav** [comment]
@chiragsalian Any updates for me on this one? As I've mentioned earlier I'll resolve conflicts once all review is complete or when there are changes. I keep getting conflicts because I've updated the function component to class component.

**Comment by chiragsalian** [comment]
Sorry yes, I was out for Thanksgiving week. Checking this now.

**Comment by chiragsalian** [comment]
> It seems to be happening due to Navigation.dismissModal() in Policy.js:navigateToPolicy method. If I remove the dismissModal call, it works fine. I am not sure why it was added. Can you confirm?

My guess is because earlier we use to show a large full screen modal and when we clicked new workspace there we wanted the modal to close. But we don't show that full-screen modal anymore so i'm pretty sure you are safe to remove it. Its usages are also only in two spots so it should be easy enough to confirm if everything runs fine.

> I had this issue when I raised the first draft, but not later when I submitted with screenshots. Tried creating multiple workspaces and it didn't end into a loop. Can you take the latest pull and check?

Cool on investigation it looks like its not because of your PR and the issue lies elsewhere.

**Review by chiragsalian — DISMISSED** [review]
Changes LGTM. Feel free to address the conflicts and remove `Navigation.dismissModal` (test it a bit too)
and then i can merge your PR 👍

**Comment by mananjadhav** [comment]
> Sorry yes, I was out for Thanksgiving week. Checking this now.

@chiragsalian  No problem. Thanks for the reviewing, PR updated.

**Review by chiragsalian — APPROVED** [review]
LGTM and works well. Thank you for the changes.

**Comment by marcaaron** [comment]
>My guess is because earlier we use to show a large full screen modal and when we clicked new workspace there we wanted the modal to close. But we don't show that full-screen modal anymore so i'm pretty sure you are safe to remove it. Its usages are also only in two spots so it should be easy enough to confirm if everything runs fine.

It was actually because when creating a new workspace via the `LogInWithShortLivedTokenPage` we must dismiss the modal that shows a loading spinner here:

https://github.com/Expensify/App/blob/6433741e0080f9767602829871c1ea34d4ec6de3/src/pages/LogInWithShortLivedTokenPage.js#L92

And the removal of the call to `dismissModal()` broke this is not obvious ways e.g.

https://github.com/Expensify/Expensify/issues/188048

This was probably hard (maybe impossible) to see because the line from `LoginWithShortLivedTokenPage` to policy creation (and which modal needs to be dismissed) isn't super clear.

cc @roryabraham @ctkochan22

**Comment by mananjadhav** [comment]
@marcaaron Then should I add a param to the function and add the `Navigation.dismissModal` back.

**Comment by marcaaron** [comment]
No worries @mananjadhav we'll take care of it.

**Comment by mananjadhav** [comment]
Thanks, @marcaaron. Do let me know if any help is required.

**Pull request merged**